File: problems/Add Binary/addBinary.cs
public class Solution {
    public string AddBinary(string a, string b) {
        var result = new Stack<int>();
        var memory = 0;
        var aIndex = a.Length - 1;
        var bIndex = b.Length - 1;

        while (0 <= aIndex || 0 <= bIndex || 1 == memory) {
            var aDigit = 0 <= aIndex && '1' == a[aIndex--] ? 1 : 0;
            var bDigit = 0 <= bIndex && '1' == b[bIndex--] ? 1 : 0;
            result.Push(aDigit ^ bDigit ^ memory);
            memory = (1 == (aDigit & bDigit)) || (1 == ((aDigit ^ bDigit) & memory)) ? 1 : 0;
        }

        return string.Join("", result);
    }
}

File: problems/Add Binary/addBinary.java
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int aIdx = a.length() - 1;
        int bIdx = b.length() - 1;
        int memory = 0;

        while (0 <= aIdx || 0 <= bIdx || 1 == memory) {
            int digit = memory;
            digit += (0 <= aIdx && '1' == a.charAt(aIdx--) ? 1 : 0);
            digit += (0 <= bIdx && '1' == b.charAt(bIdx--) ? 1 : 0);

            sb.insert(0, digit % 2);
            memory = digit >> 1;
        }

        return sb.toString();
    }
}
